# large_scale_genomic_alignment_extraction
A scalable large scale genomic fraction aligner and extractor that uses numba for the large scale alignment of the genomes and the transcriptomes and process them over the cores for the extraction of the aligned regions. The aligned regions can also be mapped to the length plotter and can be machine trained for specific applications. This repository contains two codes one for the one-one genome alignment and extractions and it is used for one reference genome and one target genome and the other code in which you can have one target genome and as many as reference genomes you want. 
# in case of the one genome one target, such as bacterial or the organell genome alignments, it provides the target genome start, end, reference genome start end and also corrdinates and reverse corrdinates and sequences. \
# in case of the one 2 many genome, such as a specific transcript alignment to the larger database of the transcripts, or genome, it provides, the target start, end, cordinates, sequences and returns a list of lists with the named tuple as the referencename, sequenceextracted. \
Gaurav Sablok \
Senior Postdoctoral Fellow \
Faculty of Natural and Agricultural Sciences Room 7-35, \
Agricultural Sciences Building University of Pretoria, \
Private Bag X20 Hatfield 0028, South Africa \
Academia : https://up-za.academia.edu/GauravSablok \
Frontiers: https://loop.frontiersin.org/people/33293/overview \
ORCID: https://orcid.org/0000-0002-4157-9405 \
WOS: https://www.webofscience.com/wos/author/record/C-5940-2014 \
Github:https://github.com/sablokgaurav \
Linkedln: https://www.linkedin.com/in/sablokgaurav/ \
ResearchGate: https://www.researchgate.net/profile/Gaurav-Sablok \
Plant Bioinformatics | Illumina | PacBio | OxfordNanoPore | Transcriptomics | \
Genomics | Metagenomics | DevOPS | Data Engineering and Analyst | \
Python | R | Java | Scala | Bash | DJANGO | Machine Learning \
